dashboard/modules/job/job_manager.py

```diff
@@ -315,8 +317,11 @@ async def _polling(self, child_process) -> int:
                     await asyncio.sleep(self.SUBPROCESS_POLL_PERIOD_S)
         except Exception:
             if child_process:
-                # TODO (jiaodong): Improve this with SIGTERM then SIGKILL
-                child_process.kill()
+                os.killpg(os.getpgid(child_process.pid), signal.SIGTERM)
+                try:
+                    child_process.wait(self.WAIT_FOR_JOB_TERMINATION_S)
```

Comment:
According to https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait

> The function is implemented using a busy loop (non-blocking call and short sleeps). Use the [asyncio](https://docs.python.org/3/library/asyncio.html#module-asyncio) module for an asynchronous wait: see [asyncio.create_subprocess_exec](https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.create_subprocess_exec).

Since we're in an `async` function, should we use the asynchronous wait? (I'm not actually sure what they mean by "non-blocking call and short sleeps" here, should I think of it as being as bad as `time.sleep()`? Maybe you or other reviewers know more.)

Reply:
yes, try https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor

Reply:
@architkulkarni Changed to use asynchronous polling, PTAL!